File: src/euler_003.jl
is_prime(n) = all(x -> n % x != 0, 2:isqrt(n))

function answer(n)
    filter(x -> (n % x == 0 && is_prime(x)), 3:2:isqrt(n)) |> last
end

answer(600851475143) |> print

File: src/euler_003.hs
isqrt :: Int -> Int
isqrt = floor . sqrt . fromIntegral

factors :: Int -> [Int]
factors n = [x | x <- [2..isqrt n], n `mod` x == 0]

isPrime :: (Int) -> Bool
isPrime n = factors n == []

answer :: Int -> Int
answer x = maximum $ filter (isPrime) (factors x)

main = putStrLn $ show $ answer 600851475143
